src/contracts/GPv2Settlement.sol

```diff
@@ -54,188 +31,51 @@ contract GPv2Settlement {
-    /// to reduce the amount of call data required to call this method. Orders
-    /// encode the following fields:
-    /// ```
-    /// struct Order {
-    ///     sellAmount:     uint112,
-    ///     buyAmount:      uint112,
-    ///     validTo:        uint32,
-    ///     tip:            uint112,
-    ///     flags:          uint8,
-    ///     executedAmount: uint112,
-    ///     signature: {
-    ///         v:          uint8,
-    ///         r:          bytes32,
-    ///         s:          bytes32,
-    ///     }
-    /// }
-    /// ```
-    ///
-    /// Note that the encoded order data does not contain which token is the
-    /// sell or buy token. This data is implicit from which parameter the order
-    /// was specified from, either `encodedOrder0` (which sells token 0 and buys
-    /// token 1) or `encodedOrder1` (which sells token 1 and buys token 0).
-    /// Additionally, the nonce is not encoded as part of the order bytes as the
-    /// nonce is included in the signature. For an order to be valid it must
-    /// either be the current nonce, or be a replayable order (nonce of 0). If
-    /// the nonce were incorrect (i.e. attempted to use an order in a different
-    /// batch than it was signed for), then the signature recovery would fail.
-    /// This allows us to encode the nonce value in a single bit as part of the
-    /// flags, were a `0` represents the nonce should be for the current batch
-    /// and a `1` represents that it is a replayble order and the nonce should
-    /// be 0.
+    /// Note that some parameters are encoded as packed bytes in order to save
+    /// calldata gas. For more information on encoding format consult the
+    /// [`GPv2Encoding`] library.
     ///
-    /// @param token0 The address of token 0 being traded in the batch.
-    /// @param token1 The address of token 1 being traded in the batch.
-    /// @param d0 The amount of token 0 being traded to the AMM.
-    /// @param d1 The amount of token 1 being traded to the AMM.
-    /// @param clearingPrice0 The price of token 0 expressed in arbitrary units.
-    /// Note the exchange rate between token 0 and token 1 is `price1 / price0`.
-    /// @param clearingPrice1 The price of token 1 expressed in arbitrary units.
-    /// Note the exchange rate between token 1 and token 0 is `price0 / price1`.
-    /// @param encodedOrders0 All orders trading token 0 for token 1, that is,
-    /// either orders selling token 0 for token 1 or buying token 1 for token 0.
-    /// @param encodedOrders1 All orders trading token 1 for token 2.
+    /// @param tokens An array of ERC20 tokens to be traded in the settlement.
+    /// Orders and interactions encode tokens as indices into this array.
+    /// @param clearingPrices An array of clearing prices where the `i`-th price
+    /// is for the `i`-th token in the [`tokens`] array.
+    /// @param feeFactor The fee factor to use for the trade. The actual fee
```

Comment:
Is there some more explanation on the design decision of passing in a variable fee per batch instead of using a hardcoded fee? It would be great to see an example use case where this is beneficial.

Reply:
Its useful for settling a user order with Uniswap directly for example.

Reply:
In this example wouldn't we be able to use the strategy we discussed before, where we pick the clearing price at _uniswapEffectivePrice made better by 0.1%_? This way the user effectively pays the uniswap price (after fees), and since the settlement contract would itself also only be able to get _uniswapEffectivePrice_ from Uniswap, the fees collected in the end are effectively 0.

I was thinking more of an example where it's beneficial because of some buffering (which is an example Alex alluded to)

Reply:
I think for buffering, it is always the best to use the maximal fee, as it means a maximal buffer.

But for all sorts of promotions or beating competition, it is nice to be able to lower the fees.